### PR TITLE
Stats: Fix purchase recoganization on Odyssey

### PR DESCRIPTION
**Pull request by kangzj**
Related to #

## Proposed Changes

* update product and purchase fetching for Odyssey Stats to include Stats plans with tiers

## Testing Instructions

* setup Odyssey locally according to a FG guide
* make sure that you blog doesn't have any Stats plans
* purchase a plan and verify that you are redirected to the Traffic page instead of the purchase page again

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: apps/odyssey-stats/src/components/odyssey-query-products.ts [hunk 1]
```diff
@@ -1,8 +1,31 @@
...
+				new URLSearchParams( { currency: currency ?? '', type } )
+		)
+		.then( ( response ) => response.json() )
+		.catch( ( error: Error ) => error );
+}
+function useQueryProductsList( currency = '', type = 'jetpack' ) {
+	return useQuery( {
+		queryKey: [ 'odyssey-stats', 'products', currency, type ],
+		queryFn: () => queryProductsList( currency, type ),
+		staleTime: 10 * 1000,
```
Comment (kangzj) [inline apps/odyssey-stats/src/components/odyssey-query-products.ts:22]: Basically the stale time saves the day!

File: apps/odyssey-stats/src/components/odyssey-query-site-purchases.ts [hunk 2]
```diff
@@ -1,59 +1,75 @@
...
 /**
  * Update site products in the Redux store by fetching purchases via API for Odyssey Stats.
  */
+
 const useOdysseyQuerySitePurchases = ( siteId: number | null ) => {
-	const isRequesting = useSelector( isFetchingSitePurchases );
+	return useQuery( {
+		queryKey: [ 'odyssey-stats', 'site-purchases', siteId ],
+		queryFn: () => queryOdysseyQuerySitePurchases( siteId ),
+		staleTime: 10 * 1000,
```
Comment (kangzj) [inline apps/odyssey-stats/src/components/odyssey-query-site-purchases.ts:33]: Basically the stale time saves the day!

File: client/my-sites/stats/stats-redirect/index.tsx [hunk 3]
```diff
@@ -28,8 +27,13 @@ const StatsRedirectFlow: React.FC< StatsRedirectFlowProps > = ( { children } ) =
...
 	const isOdysseyStats = config.isEnabled( 'is_running_in_jetpack_site' );
 
-	const { isFreeOwned, isPWYWOwned, isCommercialOwned, supportCommercialUse } =
-		useStatsPurchases( siteId );
+	const {
+		isFreeOwned,
+		isPWYWOwned,
+		isCommercialOwned,
+		supportCommercialUse,
+		isRequestingSitePurchases,
```
Comment (kangzj) [inline client/my-sites/stats/stats-redirect/index.tsx:35]: We need to wait until site purchases request is finished

File: client/my-sites/stats/stats-redirect/index.tsx [hunk 4]
```diff
@@ -60,7 +65,6 @@ const StatsRedirectFlow: React.FC< StatsRedirectFlowProps > = ( { children } ) =
 		purchaseNotPosponed &&
 		qualifiedUser;
 
-	const isRequesting = useSelector( ( state: object ) => isStatsNoticeSettingsFetching( state ) );
```
Comment (kangzj) [inline client/my-sites/stats/stats-redirect/index.tsx:63]: I don't think this actually does anything, so I removed it.

File: client/my-sites/stats/stats-redirect/load-stats-page.tsx [hunk 5]
```diff
@@ -99,5 +100,11 @@ export default function LoadStatsPage( { children }: AsyncLoadProps ) {
 		return <EnableStatsModuleNotice siteId={ siteId } path={ path } />;
 	}
 
-	return <StatsRedirectFlow>{ children }</StatsRedirectFlow>;
+	return (
+		<StatsRedirectFlow>
+			{ /* Only query site purchases on Calypso via existing data component */ }
+			<QuerySitePurchases siteId={ siteId } />
```
Comment (kangzj) [inline client/my-sites/stats/stats-redirect/load-stats-page.tsx]: We need site purchase for the flow

File: client/my-sites/stats/stats-redirect/load-stats-page.tsx [hunk 6]
```diff
@@ -99,5 +100,11 @@ export default function LoadStatsPage( { children }: AsyncLoadProps ) {
 		return <EnableStatsModuleNotice siteId={ siteId } path={ path } />;
 	}
 
-	return <StatsRedirectFlow>{ children }</StatsRedirectFlow>;
+	return (
+		<StatsRedirectFlow>
+			{ /* Only query site purchases on Calypso via existing data component */ }
```
Comment (kangzj) [inline client/my-sites/stats/stats-redirect/load-stats-page.tsx]:
```suggestion
```

File: client/my-sites/stats/stats-redirect/load-stats-page.tsx [hunk 7]
```diff
@@ -99,5 +100,10 @@ export default function LoadStatsPage( { children }: AsyncLoadProps ) {
 		return <EnableStatsModuleNotice siteId={ siteId } path={ path } />;
 	}
 
-	return <StatsRedirectFlow>{ children }</StatsRedirectFlow>;
+	return (
+		<StatsRedirectFlow>
+			{ isOdysseyStats && <QuerySitePurchases siteId={ siteId } /> }
```
Comment (kangzj) [inline client/my-sites/stats/stats-redirect/load-stats-page.tsx:105]: hmmm it seems like the query is already sent by Calypso somewhere, hence the dead loop for Calypso Stats... 😢

**Review by grzegorz-cp — APPROVED** [review]

**Review by kangzj — COMMENTED** [review]
It tests well for me. Thanks for fixing it. Nice one! @grzegorz-cp 👍 

PS: You theory makes sense that the query component might have been destroyed after the redirection.

**Pull request merged**